drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,309 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Minimal driver for FMC-attached NOR flash, implementing the `hf` API
+//!
+//! The NOR flash chip is a W25Q01JVZEIQ, which is a 1 GiB NOR flash.  It is
+//! connected to the FPGA over SPI / QSPI.
+//!
+//! # References
+//! - [SPI NOR controller](https://github.com/oxidecomputer/quartz/tree/main/hdl/ip/vhd/spi_nor_controller/docs)
+
+#![no_std]
+#![no_main]
+
+use userlib::hl::sleep_for;
+
+use ringbuf::{counted_ringbuf, ringbuf_entry};
+
+mod hf; // implementation of `HostFlash` API
+
+#[derive(Debug, Clone, Copy, PartialEq, counters::Count)]
+enum Trace {
+    #[count(skip)]
+    None,
+
+    FpgaBusy(u32),
+    SectorEraseBusy,
+    WriteBusy,
+
+    HashUpdateError(drv_hash_api::HashError),
+    HashFinalizeError(drv_hash_api::HashError),
+}
+
+counted_ringbuf!(Trace, 32, Trace::None);
+
+/// Size in bytes of a single page of data (i.e., the max length of slice we
+/// accept for `page_program()` and `read_memory()`).
+///
+/// This value is really a property of the flash we're talking to and not this
+/// driver, but it's correct for all our current parts. If that changes, this
+/// will need to change to something more flexible.
+pub const PAGE_SIZE_BYTES: usize = 256;
+
+/// Size in bytes of a single sector of data (i.e., the size of the data erased
+/// by a call to `sector_erase()`).
+///
+/// This value is really a property of the flash we're talking to and not this
+/// driver, but it's correct for all our current parts. If that changes, this
+/// will need to change to something more flexible.
+///
+/// **Note:** the datasheet refers to a "sector" as a 4K block, but also
+/// supports 64K block erases, so we call the latter a sector to match the
+/// behavior of the Gimlet host flash driver.
+pub const SECTOR_SIZE_BYTES: u32 = 65_536;
+
+#[export_name = "main"]
+fn main() -> ! {
+    // Wait for the FMC to be configured
+    userlib::hl::sleep_for(1000);
```

Comment:
how do we know that it will take less than 1000 ticks to do this? What happens if we sleep and the FMC still hasn't been configured?

Reply:
> What happens if we sleep and the FMC still hasn't been configured?

Panic! at the ID check

(I added a basic `ping` function to the sequencer API, so `cosmo-hf` now waits for that to return before starting up)